File: architecture-skills/object-oriented-programming-paradigm/examples/classes-and-objects/classes-and-objects.cs
// Classes & objects in C#.
//
// C# is class-mandatory like Java, so transfer() lives as a static
// method. Properties (Balance { get; private set; }) replace public
// fields, and a static member holds the per-class account counter.
// Top-level statements form the demo entry point.

using System;

BankAccount alice = new("Alice", 100.00);
BankAccount bob = new("Bob");

alice.Deposit(50.00);
bob.Deposit(20.00);
alice.Withdraw(30.00);

Console.WriteLine(alice); // Bob's balance is untouched
Console.WriteLine(bob);

Bank.Transfer(alice, bob, 40.00);
Console.WriteLine("after transfer:");
Console.WriteLine(alice);
Console.WriteLine(bob);

Console.WriteLine($"accounts opened: {BankAccount.TotalOpened}");

public sealed class BankAccount
{
    // Static member: one counter shared by all instances.
    private static int accountsOpened;

    // Instance state: one copy per object, never shared.
    public string Number { get; }
    public string Owner { get; }
    public double Balance { get; private set; }

    // Constructor: the only way in; if it returns, state is valid.
    public BankAccount(string owner, double openingBalance = 0.0)
    {
        if (openingBalance < 0)
            throw new ArgumentOutOfRangeException(nameof(openingBalance));
        Number = $"ACC-{++accountsOpened:D4}";
        Owner = owner;
        Balance = openingBalance;
    }

    // Instance methods: implicit `this` binds them to one object.
    public void Deposit(double amount)
    {
        if (amount <= 0)
            throw new ArgumentOutOfRangeException(nameof(amount));
        Balance += amount;
    }

    public void Withdraw(double amount)
    {
        if (amount > Balance)
            throw new InvalidOperationException($"{Number}: insufficient funds");
        Balance -= amount;
    }

    // Static property: bound to the class, readable with no instance.
    public static int TotalOpened => accountsOpened;

    public override string ToString() => $"{Number} ({Owner}): ${Balance:F2}";
}

// No free functions in C#: a static method on a utility class is the
// closest fit for behavior coordinating two objects that owns neither.
public static class Bank
{
    public static void Transfer(BankAccount from, BankAccount to, double amount)
    {
        from.Withdraw(amount);
        to.Deposit(amount);
    }
}

File: architecture-skills/object-oriented-programming-paradigm/examples/classes-and-objects/classes-and-objects.js
// Classes & objects in JavaScript.
//
// `class` is sugar over prototypes: methods live once on the
// prototype, fields live on each instance. #private fields keep the
// balance inaccessible from outside, `static` members hang off the
// class itself, and transfer() stays a plain function because it
// belongs to no single account.

class BankAccount {
  // Static field: one counter shared by all instances.
  static #accountsOpened = 0;

  // Private instance fields: one copy per object, never shared.
  #balance;

  // Constructor: the only way in; if it returns, state is valid.
  constructor(owner, openingBalance = 0) {
    if (openingBalance < 0) {
      throw new RangeError("opening balance cannot be negative");
    }
    BankAccount.#accountsOpened += 1;
    this.number = `ACC-${String(BankAccount.#accountsOpened).padStart(4, "0")}`;
    this.owner = owner;
    this.#balance = openingBalance;
  }

  // Instance methods: `this` binds them to one object.
  deposit(amount) {
    if (amount <= 0) {
      throw new RangeError("deposit must be positive");
    }
    this.#balance += amount;
  }

  withdraw(amount) {
    if (amount > this.#balance) {
      throw new Error(`${this.number}: insufficient funds`);
    }
    this.#balance -= amount;
  }

  get balance() {
    return this.#balance;
  }

  toString() {
    return `${this.number} (${this.owner}): $${this.#balance.toFixed(2)}`;
  }

  // Static method: bound to the class, callable with no instance.
  static totalOpened() {
    return BankAccount.#accountsOpened;
  }
}

// Free function: coordinates two objects, owns neither.
function transfer(from, to, amount) {
  from.withdraw(amount);
  to.deposit(amount);
}

// Two objects from one blueprint, each with independent state.
const alice = new BankAccount("Alice", 100);
const bob = new BankAccount("Bob");

alice.deposit(50);
bob.deposit(20);
alice.withdraw(30);

console.log(alice.toString()); // Bob's balance is untouched
console.log(bob.toString());

transfer(alice, bob, 40);
console.log("after transfer:");
console.log(alice.toString());
console.log(bob.toString());

console.log(`accounts opened: ${BankAccount.totalOpened()}`);
